File: Kotlin/two-sum.kt
// Time:  O(n)
// Space: O(n)

class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val lookup = HashMap<Int, Int>()
        nums.forEachIndexed { i, x ->
            val j = lookup[target - x]
            j?.let {
                return intArrayOf(j, i)
            }
            lookup[x] = i
        }
        return intArrayOf()
    }
}

File: TypeScript/two-sum.ts
// Time:  O(n)
// Space: O(n)

function twoSum(nums: number[], target: number): number[] {
    let lookup = new Map();
    for (let i = 0; i < nums.length; i++) {
        if (lookup.has(target - nums[i])) {
            return [lookup.get(target - nums[i]), i];
        }
        lookup.set(nums[i], i);
    }
    return [];
};
